j5/backends/hardware/j5/serial.py

```diff
@@ -0,0 +1,125 @@
+    interface. It will catch and rethrow the errors as a CommunicationError, so that it
+    is more explicit what is going wrong.
+    """
+    @wraps(func)
+    def catch_exceptions(*args, **kwargs):  # type: ignore
+        try:
+            return func(*args, **kwargs)
+        except SerialTimeoutException as e:
+            raise CommunicationError(f"Serial Timeout Error: {e}")
+        except SerialException as e:
+            raise CommunicationError(f"Serial Error: {e}")
+    return catch_exceptions
+
+
+class Seriallike(Protocol):
+    """
+    Something that walks like a Serial and quacks like a Serial.
+
+    This is used instead of hardcoding the Serial class to allow it to be mocked out.
+    """
+
+    def __init__(self,
+                 port: Optional[str] = None,
+                 baudrate: int = 9600,
+                 bytesize: int = 8,
+                 parity: str = 'N',
+                 stopbits: float = 1,
+                 timeout: Optional[float] = None):
+        ...
+
+    def close(self) -> None:
+        """Close the connection."""
+        ...
+
+    def flush(self) -> None:
+        """Flush all pending write operations."""
+        ...
+
+    def readline(self) -> bytes:
+        """Read a line from the serial port."""
+        ...
+
+    def write(self, data: bytes) -> int:
+        """Write data to the serial port."""
+        ...
+
+
+class SerialHardwareBackend(metaclass=BackendMeta):
+    """An abstract class for creating backends that use USB serial communication."""
+
+    @handle_serial_error
+    def __init__(
+            self,
+            serial_port: str,
+            serial_class: Type[Seriallike] = Serial,
+            baud: int = 115200,
+            timeout: timedelta = timedelta(milliseconds=250),
+    ) -> None:
+        timeout_secs = timeout / timedelta(seconds=1)
+        self._serial = serial_class(
+            port=serial_port,
+            baudrate=baud,
+            timeout=timeout_secs,
+        )
+
+    @classmethod
+    @abstractmethod
+    def discover(cls) -> Set[Board]:
+        """Discover boards that this backend can control."""
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    @abstractmethod
+    def environment(self) -> Environment:
+        """Environment the backend belongs too."""
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    @abstractmethod
+    def firmware_version(self) -> Optional[str]:
+        """The firmware version of the board."""
+        raise NotImplementedError  # pragma: no cover
+
+    @handle_serial_error
+    def read_serial_line(self, empty: bool = False) -> str:
+        """Read a line from the serial interface."""
+        bdata = self._serial.readline()
+
+        if len(bdata) == 0:
+            if empty:
+                return ""
+            raise CommunicationError(
+                "No response from board. "
```

Comment:
We should be able to add `board.name` here to give a nicer message

Reply:
Unfortunately `SerialHardwareBackend` doesn't have a `board` attribute because it is generalised across many types of board. Perhaps we should define `SerialHardwareBackend.board` to be an "abstract attribute" in the same manner as in #342.